apps/meteor/app/livechat/server/lib/contacts/migrateVisitorToContactId.ts

```diff
@@ -0,0 +1,51 @@
+import type { ILivechatVisitor, IOmnichannelSource, ILivechatContact, IOmnichannelRoom } from '@rocket.chat/core-typings';
+import { LivechatContacts, LivechatRooms } from '@rocket.chat/models';
+
+import { Livechat } from '../LivechatTyped';
+import { ContactMerger } from './ContactMerger';
+import { createContactFromVisitor } from './createContactFromVisitor';
+
+/**
+	This function assumes you already ensured that the visitor is not yet linked to any contact
+**/
+export async function migrateVisitorToContactId(
+	visitor: ILivechatVisitor,
+	source: IOmnichannelSource,
+	requireRoom = true,
+): Promise<ILivechatContact['_id'] | null> {
+	if (requireRoom) {
+		// Do not migrate the visitor with this source if they have no rooms matching it
+		const anyRoom = await LivechatRooms.findNewestByContactVisitorAssociation<Pick<IOmnichannelRoom, '_id'>>(
+			{ visitorId: visitor._id, source },
+			{
+				projection: { _id: 1 },
+			},
+		);
+
+		if (!anyRoom) {
+			return null;
+		}
+	}
+
+	// Search for any contact that is not yet associated with any visitor and that have the same email or phone number as this visitor.
+	const existingContact = await LivechatContacts.findContactMatchingVisitor(visitor);
+	if (!existingContact) {
+		Livechat.logger.debug(`Creating a new contact for existing visitor ${visitor._id}`);
```

Comment:
Same for these, no need to get LivechatTyped. There should be a logger with the same props on `server/lib/logger.ts` file

Reply:
This should be fixed now!